### PR TITLE
[ISSUE #139]add failfast. resolve #139

### LINKED ISSUE #139: add failfastCluster
**What would you like to be added**:
add failfastCluster

**Why is this needed**:
be consistent with java version

### PR DESCRIPTION
**Pull request by xujianhai666**
**What this PR does**:
add failfastClusterInvoker

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cluster/cluster_impl/base_cluster_invoker.go [hunk 1]
```diff
@@ -18,6 +18,8 @@
 package cluster_impl
 
 import (
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/extension"
 	perrors "github.com/pkg/errors"
 	"go.uber.org/atomic"
 )
```
Comment (hxmhlt) [inline cluster/cluster_impl/base_cluster_invoker.go]: import format

File: cluster/cluster_impl/failover_cluster_invoker.go [hunk 2]
```diff
@@ -48,17 +47,11 @@ func (invoker *failoverClusterInvoker) Invoke(invocation protocol.Invocation) pr
...
-	lb := url.GetParam(constant.LOADBALANCE_KEY, constant.DEFAULT_LOADBALANCE)
+	loadbalance := getLoadBalance(invokers[0], invocation)
 
-	//Get the service method loadbalance config if have
-	if v := url.GetMethodParam(methodName, constant.LOADBALANCE_KEY, ""); v != "" {
-		lb = v
-	}
-	loadbalance := extension.GetLoadbalance(lb)
+	methodName := invocation.MethodName()
+	url := invoker.GetUrl()
```
Comment (hxmhlt) [inline cluster/cluster_impl/failover_cluster_invoker.go]: not invoker.GetUrl() ,but invokers[0].
invoker.GetUrl() will return directory.url , it is registry's url if use registry mode.

File: cluster/cluster_impl/base_cluster_invoker.go (hunk 1 above)
Comment (xujianhai666) [reply cluster/cluster_impl/base_cluster_invoker.go]: done

File: cluster/cluster_impl/failover_cluster_invoker.go (hunk 2 above)
Comment (xujianhai666) [reply cluster/cluster_impl/failover_cluster_invoker.go]: done

File: cluster/cluster_impl/failfast_cluster_test.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/apache/dubbo-go/protocol"
+	"github.com/apache/dubbo-go/protocol/invocation"
+)
+
+var (
+	failfastUrl, _ = common.NewURL(context.TODO(), "dubbo://192.168.1.1:20000/com.ikurento.user.UserProvider")
+)
+
+// register_failfast register failfastCluster to cluster extension.
+func register_failfast(t *testing.T, invoker *protocol.MockInvoker) protocol.Invoker {
```
Comment (AlexStocks) [inline cluster/cluster_impl/failfast_cluster_test.go]: registerFailfast

File: cluster/cluster_impl/failfast_cluster_invoker.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	return &failfastClusterInvoker{
+		baseClusterInvoker: newBaseClusterInvoker(directory),
+	}
+}
+
+func (invoker *failfastClusterInvoker) Invoke(invocation protocol.Invocation) protocol.Result {
+
+	invokers := invoker.directory.List(invocation)
+	err := invoker.checkInvokers(invokers, invocation)
+
```
Comment (AlexStocks) [inline cluster/cluster_impl/failfast_cluster_invoker.go]: delete this line

File: cluster/cluster_impl/base_cluster_invoker.go [hunk 5]
```diff
@@ -115,12 +117,24 @@ func (invoker *baseClusterInvoker) doSelect(lb cluster.LoadBalance, invocation p
...
+
+func getLoadBalance(invoker protocol.Invoker, invocation protocol.Invocation) cluster.LoadBalance {
+	url := invoker.GetUrl()
+
+	methodName := invocation.MethodName()
+	//Get the service loadbalance config
+	lb := url.GetParam(constant.LOADBALANCE_KEY, constant.DEFAULT_LOADBALANCE)
+
+	//Get the service method loadbalance config if have
+	if v := url.GetMethodParam(methodName, constant.LOADBALANCE_KEY, ""); v != "" {
```
Comment (AlexStocks) [inline cluster/cluster_impl/base_cluster_invoker.go]: the check cond len(v) == 0 may be better

File: protocol/mock_invoker.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
+// Code generated by MockGen. DO NOT EDIT.
+// Source: invoker.go
+
+// Package protocol is a generated GoMock package.
+package protocol
+
+import (
+	common "github.com/apache/dubbo-go/common"
+	gomock "github.com/golang/mock/gomock"
+	reflect "reflect"
```
Comment (AlexStocks) [inline protocol/mock_invoker.go]: hey, guy, another import block pls.

File: protocol/mock_invoker.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
+// Code generated by MockGen. DO NOT EDIT.
```
Comment (fangyincheng) [inline protocol/mock_invoker.go]: `mock_invoker.go` should not be in directory `dubbo-go/protocol/`.  It maybe better to move it to `dubbo-go/protocol/mock`.

File: cluster/cluster_impl/failfast_cluster_test.go (hunk 3 above)
Comment (xujianhai666) [reply cluster/cluster_impl/failfast_cluster_test.go]: done

Comment (xujianhai666) [reply cluster/cluster_impl/failfast_cluster_test.go]: @AlexStocks

File: protocol/mock_invoker.go (hunk 6 above)
Comment (xujianhai666) [reply protocol/mock_invoker.go]: done @AlexStocks

File: protocol/mock_invoker.go (hunk 7 above)
Comment (xujianhai666) [reply protocol/mock_invoker.go]: done @fangyincheng

File: cluster/cluster_impl/failfast_cluster_invoker.go (hunk 4 above)
Comment (xujianhai666) [reply cluster/cluster_impl/failfast_cluster_invoker.go]: why?

File: cluster/cluster_impl/base_cluster_invoker.go (hunk 5 above)
Comment (xujianhai666) [reply cluster/cluster_impl/base_cluster_invoker.go]: done. change info len(v) > 0

File: cluster/cluster_impl/failfast_cluster_invoker.go (hunk 4 above)
Comment (xujianhai666) [reply cluster/cluster_impl/failfast_cluster_invoker.go]: done. @AlexStocks

**Pull request merged**